crates/red_knot_module_resolver/src/resolver.rs

```diff
@@ -127,83 +125,101 @@ pub struct ModuleResolutionSettings {
-    ///
-    /// (TODO: fall back to vendored stubs if no custom directory is provided.)
-    pub custom_typeshed: Option<FileSystemPathBuf>,
 }
 
 impl ModuleResolutionSettings {
     /// Implementation of PEP 561's module resolution order
     /// (with some small, deliberate, differences)
-    fn into_ordered_search_paths(self) -> OrderedSearchPaths {
+    ///
+    /// TODO(Alex): this method does multiple `.unwrap()` calls when it should really return an error.
+    /// Each `.unwrap()` call is a point where we're validating a setting that the user would pass
+    /// and transforming it into an internal representation for a validated path.
+    /// Rather than panicking if a path fails to validate, we should display an error message to the user
+    /// and exit the process with a nonzero exit code.
+    /// This validation should probably be done outside of Salsa?
+    fn into_resolved_settings(self) -> ResolvedModuleResolutionSettings {
         let ModuleResolutionSettings {
+            target_version,
             extra_paths,
             workspace_root,
             site_packages,
             custom_typeshed,
         } = self;
 
-        let mut paths: Vec<_> = extra_paths
+        let mut paths: Vec<ModuleResolutionPathBuf> = extra_paths
             .into_iter()
-            .map(|path| ModuleSearchPath::new(path, ModuleSearchPathKind::Extra))
+            .map(|fs_path| ModuleResolutionPathBuf::extra(fs_path).unwrap())
             .collect();
 
-        paths.push(ModuleSearchPath::new(
-            workspace_root,
-            ModuleSearchPathKind::FirstParty,
-        ));
+        paths.push(ModuleResolutionPathBuf::first_party(workspace_root).unwrap());
 
-        // TODO fallback to vendored typeshed stubs if no custom typeshed directory is provided by the user
         if let Some(custom_typeshed) = custom_typeshed {
-            paths.push(ModuleSearchPath::new(
-                custom_typeshed.join(TYPESHED_STDLIB_DIRECTORY),
-                ModuleSearchPathKind::StandardLibrary,
-            ));
+            paths.push(
+                ModuleResolutionPathBuf::stdlib_from_typeshed_root(&custom_typeshed).unwrap(),
+            );
         }
 
         // TODO vendor typeshed's third-party stubs as well as the stdlib and fallback to them as a final step
         if let Some(site_packages) = site_packages {
-            paths.push(ModuleSearchPath::new(
-                site_packages,
-                ModuleSearchPathKind::SitePackagesThirdParty,
-            ));
+            paths.push(ModuleResolutionPathBuf::site_packages(site_packages).unwrap());
         }
 
-        OrderedSearchPaths(paths)
+        ResolvedModuleResolutionSettings {
+            target_version,
+            search_paths: OrderedSearchPaths(paths.into_iter().map(Arc::new).collect()),
+        }
     }
 }
 
 /// A resolved module resolution order, implementing PEP 561
 /// (with some small, deliberate differences)
```

Comment:
Should this comment reference the appropriate section of the typing spec (which is what we actually care about) rather than the PEP?

Do we have any remaining intentional differences with the typing spec?

Reply:
Yes, this comment is now out of date; good catch.